evm/db/diff.py

```diff
@@ -0,0 +1,113 @@
+    Union,
+)
+
+from cytoolz import (
+    merge,
+)
+
+
+class MissingReason:
+    def __init__(self, reason):
+        self.reason = reason
+
+    def __str__(self, reason):
+        return "Key is missing because it was {}".format(self.reason)
+
+
+NEVER_INSERTED = MissingReason("never inserted")
+DELETED = MissingReason("deleted")
+
+
+class DiffMissingError(KeyError):
+    def __init__(self, missing_key: bytes, reason: MissingReason) -> None:
+        self.reason = reason
+        super().__init__(missing_key, reason)
+
+    def is_deleted(self):
+        return self.reason == DELETED
+
+
+class DBDiffTracker(MutableMapping):
+    """
+    Recorded changes to a :class:`~evm.db.BaseDB`
+    """
+    def __init__(self):
+        self._changes = {}  # type: Dict[bytes, Union[bytes, DiffMissingError]]
+
+    def __contains__(self, key):
+        result = self._changes.get(key, NEVER_INSERTED)
+        return result not in (DELETED, NEVER_INSERTED)
+
+    def __getitem__(self, key):
+        result = self._changes.get(key, NEVER_INSERTED)
+        if result in (DELETED, NEVER_INSERTED):
+            raise DiffMissingError(key, result)
+        else:
+            return result
+
+    def __setitem__(self, key, value):
+        self._changes[key] = value
+
+    def __delitem__(self, key):
+        # The diff does not have access to any underlying db,
+        # so it cannot check if the key exists before deleting.
+        self._changes[key] = DELETED
+
+    def __iter__(self):
+        raise NotImplementedError(
+            "Cannot iterate through changes, use diff().apply_to(db) to update a database"
+        )
+
+    def __len__(self):
+        return len(self._changes)
+
+    def diff(self):
+        return DBDiff(self)
+
+
+class DBDiff(Mapping):
+    _changes = None  # type: Dict[bytes, Union[bytes, DiffMissingError]]
+
+    def __init__(self, tracker: DBDiffTracker = None) -> None:
+        if tracker is None:
+            self._changes = {}
+        else:
+            self._changes = tracker._changes
+
+    def __getitem__(self, key):
+        result = self._changes.get(key, NEVER_INSERTED)
+        if result in (DELETED, NEVER_INSERTED):
+            raise DiffMissingError(key, result)
+        else:
+            return result
+
+    def __iter__(self):
+        raise NotImplementedError(
+            "Cannot iterate through changes, use apply_to(db) to update a database"
+        )
+
+    def __len__(self):
+        return len(self._changes)
+
+    def apply_to(self, db, apply_deletes=True):
+        for key, value in self._changes.items():
+            if value is DELETED and apply_deletes:
+                try:
+                    del db[key]
+                except KeyError:
+                    pass
+            else:
+                db[key] = value
+
+    @classmethod
+    def join(cls, diffs: Iterable['DBDiff']) -> 'DBDiff':
+        new_diff = cls()
+        new_diff._changes = merge(diff._changes for diff in diffs)
```

Comment:
This private attribute access makes me thing that the class should be able to be instantiated with an existing dictionary of diff changes.

Reply:
I wanted to avoid a public way to initialize with a `dict` of changes, because it would be easy to do it wrong (like assume that a `None`) value would delete the underlying. These two classes inherently share this special type of dict.

So I guess a third option is to create a third class whose job is essentially to document the deletion mechanism, and do a little type checking with mypy, like:

```py
class DBDiffChanges(dict):
    """
    Changes recorded by `DBDiffTracker` and later used by `DBDiff`. It
    has a special internal mechanism for tracking deletes. Best not to try
    to build this manually, and to use the above two classes, instead.
    """
    pass
```

Reply:
It seems like we should be able to get pretty far with mypy checks and maybe even explicit validation of key/value types if the overhead is acceptable.

Reply:
do what you think is best